src/main/java/by/exadel/internship/service/impl/UserTimeSlotServiceImpl.java

```diff
@@ -0,0 +1,112 @@
+import by.exadel.internship.repository.UserTimeSlotRepository;
+import by.exadel.internship.service.UserTimeSlotService;
+import by.exadel.internship.service.UserService;
+import by.exadel.internship.util.MDCLog;
+import lombok.RequiredArgsConstructor;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.stereotype.Service;
+
+import java.time.Duration;
+import java.time.LocalDateTime;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.UUID;
+
+;
+
+
+@Service
+@RequiredArgsConstructor
+@Slf4j
+public class UserTimeSlotServiceImpl implements UserTimeSlotService {
+
+    private final UserTimeSlotRepository userTimeSlotRepository;
+    private final UserService userService;
+    private final UserTimeSlotMapper mapper;
+
+    private static final String SIMPLE_CLASS_NAME = UserTimeSlotService.class.getSimpleName();
+
+
+    @Override
+    public void saveUserTime(UserDTO userDTO) {
+        MDCLog.putClassNameInMDC(SIMPLE_CLASS_NAME);
+        log.info("Try to save free time list for user with uuid = {} in DB", userDTO.getId());
+        List<UserTimeSlotWithUserDTO> resultUSerTimeList = new ArrayList<>();
+        userDTO.getUserTimeSlots().forEach(timeForCallUser -> {
+            checkTime(timeForCallUser);
+            separateTime(timeForCallUser, userDTO.getInterviewTime(), resultUSerTimeList);
+        });
+        resultUSerTimeList.forEach(time -> {
+            time.setUser(userDTO);
+        });
+        userTimeSlotRepository.saveAll(mapper.mapToEntity(resultUSerTimeList));
+        log.info("Free time list was saved in DB, user uuid = {}", userDTO.getId());
+    }
+
+    @Override
+    public void deletedById(UUID timeId) {
+        MDCLog.putClassNameInMDC(SIMPLE_CLASS_NAME);
+        log.info("Try to remove time with uuid = {} from user free time", timeId);
+        userTimeSlotRepository.deleteById(timeId);
+        log.info("Time with uuid = {} was deleted", timeId);
+    }
+
+    @Override
+    public void restoreUserTime(UserTimeSlotWithUserIdDTO time) {
+        MDCLog.putClassNameInMDC(SIMPLE_CLASS_NAME);
+        log.info("Restoring time with uuid = {} for user with uuid ={}", time.getId(), time.getUserId());
+        UserDTO userDTO = userService.getById(time.getUserId());
+        UserTimeSlotWithUserDTO userTimeSlotWithUserDTO = new UserTimeSlotWithUserDTO();
+        userTimeSlotWithUserDTO.setUser(userDTO);
+        userTimeSlotWithUserDTO.setStartDate(time.getStartDate());
+        userTimeSlotWithUserDTO.setEndDate(time.getStartDate().plusMinutes(userDTO.getInterviewTime()));
+        userTimeSlotRepository.save(mapper.toTimeForCallUserEntity(userTimeSlotWithUserDTO));
+        log.info("User time was restoring");
+    }
+
+    private void checkTime(UserTimeSlotDTO time) {
+        log.info("Bringing time to a common form");
+        time.setStartDate(commonTimeForm(time.getStartDate()));
+        time.setEndDate(commonTimeForm(time.getEndDate()));
```

Comment:
probably method `commonTimeForm` should be expanded by boolean flag (`roundUp`).
See the problem, when you round up start time, it's okay. But end time should be round down (in case round up time can be inappropriate for interviewer). 

inappropriate |  
-- | --


).

Reply:
Fixed

Reply:
Done